client/src/controllers/TabsController.ts

```diff
@@ -0,0 +1,319 @@
+      tab.removeAttribute('tabindex');
+    }
+
+    const tabContent = this.getTabPanelByHref(currentValue);
+
+    if (tabContent) {
+      if (this.animateValue) {
+        this.animateIn(tabContent);
+      } else {
+        tabContent.hidden = false;
+      }
+
+      this.dispatch('switch', {
+        detail: { tab: tab?.getAttribute('href')?.replace('#', '') },
+        target: this.listTarget,
+      });
+
+      this.dispatch('selected', {
+        cancelable: false,
+        detail: { selected: currentValue },
+      });
+
+      if (!this.syncURLHashValue) {
+        this.setURLHash(currentValue);
+      }
+    }
+  }
+
+  handleTriggerLinks(event: MouseEvent) {
+    const href = (event.target as HTMLAnchorElement).getAttribute(
+      'href',
+    ) as string;
+    const tab = this.getTabLabelByHref(href);
+    if (tab) {
+      this.selectedValue = href.replace('#', '');
+      tab.focus();
+    }
+  }
+
+  handleTabChange(event: MouseEvent) {
+    const tabId = (event.target as HTMLElement).getAttribute(
+      'aria-controls',
+    ) as string;
+    this.selectedValue = tabId;
+  }
+
+  getTabLabelByHref(tabId: string): HTMLElement | undefined {
+    return this.labelTargets.find(
+      (tab) => tab.getAttribute('aria-controls') === tabId,
+    );
+  }
+
+  getTabPanelByHref(tabId: string): HTMLElement | undefined {
+    return this.panelTargets.find((tab) => tab.getAttribute('id') === tabId);
+  }
+
+  setAriaControls(tabLinks: HTMLAnchorElement[]) {
+    tabLinks.forEach((tabLink) => {
+      const href = tabLink.getAttribute('href') as string;
+      tabLink.setAttribute('aria-controls', href.replace('#', ''));
+    });
+  }
+
+  setTabLabelIndex() {
+    (this.labelTargets as TabLink[]).forEach((label, index) => {
+      // eslint-disable-next-line no-param-reassign
+      label.index = index;
+    });
+  }
+
+  setURLHash(tabId: string) {
+    if (!window.history.state || window.history.state.tabContent !== tabId) {
+      // Add a new history item to the stack
+      window.history.pushState({ tabContent: tabId }, '', `#${tabId}`);
+    }
+  }
+
+  setTabByURLHash() {
+    if (window.location.hash) {
+      const cleanedHash = window.location.hash
+        .replace(/[^\w\-#]/g, '')
+        .replace('#', '');
+      if (cleanedHash) {
+        this.selectedValue = cleanedHash;
+      } else {
+        // The hash doesn't match a tab on the page then select first tab
+        this.selectFirstTab();
+      }
+    }
+  }
+
+  selectFirstTab() {
+    const href = this.labelTargets[0].getAttribute('aria-controls') as string;
+    this.selectedValue = href;
+  }
+
+  animateIn(tabContent: HTMLElement) {
+    debounce(() => {
+      tabContent.classList.add(...this.selectedClasses);
+    }, this.transitionValue || null)().then(() => {
+      // Wait for hidden attribute to be applied then fade in
+      // eslint-disable-next-line no-param-reassign
+      tabContent.hidden = false;
+    });
```

Comment:
```suggestion
    selectedClasses = this.selectedClasses;
    afterTransition(
      tabContent,
      // If there are no classes to add, we can skip the delay before hiding.
      selectedClasses.length ? { maxDelay: this.transitionValue } : { maxDelay: 0 },
    )(tabContent.classList.add(...selectedClasses)).then(() => {
      // eslint-disable-next-line no-param-reassign
      tabContent.hidden = false;
    });
```

This should do the same behaviour but in a more JS native API. Essentially letting the CSS run the animation and then set hidden.

We could even maybe avoid the need for transitionValue by setting a sensible default.